trl/models/utils.py

```diff
@@ -84,6 +83,17 @@ def add_hooks(model: "DeepSpeedEngine") -> None:
         optimizer_offload = model.optimizer
     else:
         raise RuntimeError("The model optimizer is None, which is not yet supported.")
+    
+    # Invalidate parameter coordinator trace to prevent stale state
+    # after generation forward passes (fixes ZeRO-3 + GKD compatibility)
+    if hasattr(optimizer_offload, "param_coordinator"):
+        coordinator = optimizer_offload.param_coordinator
+        # Only invalidate if trace is not already invalid
+        is_invalid = getattr(coordinator, "is_invalid_trace", lambda: False)()
+        if not is_invalid:
+            if hasattr(coordinator, "_invalidate_trace"):
+                coordinator._invalidate_trace()
```

Comment:
each time you use a `getattr`/`hasattr`, can you please explicitly and clearly mention via a comment in which case the object doesn't have the attribute? Eg:

```python
if not hasattr(model, "optimizer"):  # before the first training step, the model has no optimizer
```

Reply:
Thanks for the review! Updated the code.
I removed unnecessary `hasattr`/`getattr` checks for `is_invalid_trace()` and `_invalidate_trace()` since these methods always exist in `PartitionedParameterCoordinator`.
Also, I added a comment that `param_coordinator` only exists in ZeRO stage 3.